runtime-modules/content-directory/src/lib.rs

```diff
@@ -0,0 +1,1645 @@
+                }
+            });
+
+            Ok(())
+        }
+
+        pub fn remove_at_entity_property_vector(
+            origin,
+            actor: Actor<T>,
+            entity_id: T::EntityId,
+            in_class_schema_property_id: PropertyId,
+            index_in_property_vec: VecMaxLength,
+            nonce: T::Nonce
+        ) -> dispatch::Result {
+            let account_id = ensure_signed(origin)?;
+
+            Self::ensure_known_entity_id(entity_id)?;
+
+            let (entity, access_level) = Self::get_entity_and_access_level(account_id, entity_id, actor)?;
+
+            let current_property_value_vec =
+            Self::get_property_value_vec(&entity, in_class_schema_property_id)?;
+
+            Self::ensure_class_property_type_unlocked_for(
+                entity.class_id,
+                in_class_schema_property_id,
+                access_level,
+            )?;
+
+            // Ensure property value vector nonces equality to avoid possible data races,
+            // when performing vector specific operations
+            current_property_value_vec.ensure_nonce_equality(nonce)?;
+            current_property_value_vec
+                .ensure_index_in_property_vector_is_valid(index_in_property_vec)?;
+            let involved_entity_id = current_property_value_vec
+                .get_vec_value()
+                .get_involved_entities()
+                .map(|involved_entities| involved_entities[index_in_property_vec as usize]);
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            // Remove property value vector
+            <EntityById<T>>::mutate(entity_id, |entity| {
+                if let Some(PropertyValue::Vector(current_prop_value)) =
+                    entity.values.get_mut(&in_class_schema_property_id)
+                {
+                    current_prop_value.vec_remove_at(index_in_property_vec)
+                }
+            });
+            if let Some(involved_entity_id) = involved_entity_id {
+                <EntityById<T>>::mutate(involved_entity_id, |entity| entity.reference_count -= 1)
+            }
+            Ok(())
+        }
+
+        pub fn insert_at_entity_property_vector(
+            origin,
+            actor: Actor<T>,
+            entity_id: T::EntityId,
+            in_class_schema_property_id: PropertyId,
+            index_in_property_vec: VecMaxLength,
+            property_value: SinglePropertyValue<T>,
+            nonce: T::Nonce
+        ) -> dispatch::Result {
+            let account_id = ensure_signed(origin)?;
+
+            Self::ensure_known_entity_id(entity_id)?;
+
+            let (entity, access_level) = Self::get_entity_and_access_level(account_id, entity_id, actor)?;
+
+            // Try to find a current property value in the entity
+            // by matching its id to the id of a property with an updated value.
+            if let Some(PropertyValue::Vector(entity_prop_value)) =
```

Comment:
This appears to go to the mutation stage even if it's not a vector, that does not make sense, even if the mutations are safe. If you are adding an event emission in a later PR, then its a bug. I think its clearer and safer to have a proper ensure check on this.

Reply:
Good point, fixed here https://github.com/iorveth/substrate-runtime-joystream/commit/7a69ebee4f35473c27aeb449c3808cb55efd50dd